apps/04TowersOfHanoi.js

```diff
@@ -3,91 +3,133 @@
 var assert = require('assert');
 var readline = require('readline');
 var rl = readline.createInterface({
-  input: process.stdin,
-  output: process.stdout
+    input: process.stdin,
+    output: process.stdout
 });
 
 var stacks = {
-  a: [4, 3, 2, 1],
-  b: [],
-  c: []
+    a: [4, 3, 2, 1],
+    b: [],
+    c: []
 };
 
 function printStacks() {
-  console.log("a: " + stacks.a);
-  console.log("b: " + stacks.b);
-  console.log("c: " + stacks.c);
+    console.log("a: " + stacks.a);
+    console.log("b: " + stacks.b);
+    console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
+function movePiece(startStack, endStack) {
+    var block = stacks[startStack].pop();
+    stacks[endStack].push(block);
 
 }
 
-function isLegal() {
-  // Your code here
+function isLegal(startStack, endStack) {
+    var startStackArray = stacks[startStack];
+    var endStackArray = stacks[endStack];
+
+    if (startStackArray.length === 0) {
+        return false;
+    } else if (endStackArray.length === 0) {
+        return true;
+    } else {
+        return startStackArray[startStackArray.length - 1] < endStackArray[endStackArray.length - 1];
+    }
 
 }
 
 function checkForWin() {
-  // Your code here
-
+    // Your code here
+    if (stacks.b.length === 4 || stacks.c.length === 4) {
+        return true;
+    }
+    return false;
 }
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
+    // Your code here
+    if ((startStack == 'b' || startStack == 'a' || startStack == 'c') && (endStack == 'a' || endStack == 'b' || endStack == 'c')){
+        if (isLegal(startStack, endStack)) {
+            movePiece(startStack, endStack);
+        } else {
+            console.log('Not Legal');
+        }
+
+        if (checkForWin()) {
+            console.log("You Won!!!");
+            process.exit();
+        }
+
+    } else {
+        console.log("Wrong input!!!!");
+    }
```

Comment:
You're a JavaScript ninja!

![ninja](http://www.advanced-schema.com/en/img/r-d/ninja.png)